File: LeetCode_Java/binarysearch/LongestIncreasingSubsequence.java
package binarysearch;

import java.util.ArrayList;
import java.util.List;

public class LongestIncreasingSubsequence {
    public int lengthOfLIS(int[] nums) {
        List<Integer> sub = new ArrayList<>();
        sub.add(nums[0]);

        for (int i = 1; i < nums.length; i++) {
            int cur = nums[i];
            if (cur > sub.get(sub.size() - 1)) {
                sub.add(cur);
            } else {
                int index = binarySearch(sub, cur);
                sub.set(index, cur);
            }
        }

        return sub.size();
    }

    private int binarySearch(List<Integer> sub, int target) {
        int left = 0, right = sub.size();

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (sub.get(mid) < target) {
                left = mid + 1;
            } else if (sub.get(mid) > target) {
                right = mid;
            } else {
                return mid;
            }
        }

        return left;
    }

    public static void main(String[] args) {
        LongestIncreasingSubsequence solution = new LongestIncreasingSubsequence();

        int[] nums1 = {10, 9, 2, 5, 3, 7, 101, 18};
        System.out.println(solution.lengthOfLIS(nums1)); // Output: 4

        int[] nums2 = {0, 1, 0, 3, 2, 3};
        System.out.println(solution.lengthOfLIS(nums2)); // Output: 4
    }
}

File: LeetCode_Python/binarysearch/LongestIncreasingSubsequence.py
class LongestIncreasingSubsequence:
    def lengthOfLIS(self, nums: list[int]) -> int:
        res = []

        for num in nums:
            left, right = 0, len(res) - 1
            while left <= right:
                mid = (left + right) // 2
                if res[mid] < num:
                    left = mid + 1
                else:
                    right = mid - 1
            if left == len(res):
                res.append(num)
            else:
                res[left] = num

        return len(res)


if __name__ == "__main__":
    solution = LongestIncreasingSubsequence()

    print("Expected Output : 4")
    print("Actual Output : ", solution.lengthOfLIS([10, 9, 2, 5, 3, 7, 101, 18]))

    print("Expected Output : 4")
    print("Actual Output : ", solution.lengthOfLIS([0, 1, 0, 3, 2, 3]))

    print("Expected Output : 1")
    print("Actual Output : ", solution.lengthOfLIS([7, 7, 7, 7, 7, 7, 7]))
